utils/result_try.hh

```diff
@@ -0,0 +1,442 @@
+/// a try..catch block, you can use it to migrate the block so that it handles
+/// both C++ exceptions and exceptions stored in the result<>.
+///
+/// \section Example
+///
+/// Let's say that you have the following try..catch chain:
+///
+///   try {
+///       return a_function_that_may_throw();
+///   } catch (const my_exception& ex) {
+///       return 123;
+///   } catch (...) {
+///       throw;
+///   }
+///
+/// You can add support for results in the following way:
+///
+///   return utils::result_try([&] {
+///       return a_function_that_may_throw_or_return_a_failed_result();
+///   },  utils::result_catch<my_exception>([&] (const Ex&) -> result<int> {
+///       return 123;
+///   }), utils::result_catch_dots([&] (auto&& handle) -> result<int> {
+///       return handle.into_result();
+///   });
+///
+/// Each `result_catch` handler declares the exception type it handles
+/// and accepts a const reference to it as the first argument.
+/// The `result_catch_dots` handler is equivalent to `(...)` so it does not
+/// declare exception type or accept a reference to it.
+///
+/// In addition, each `result_catch` and `result_catch_dots` can optionally
+/// accept a reference to an exception handle. The exception handle can be used
+/// to return/rethrow the exception being currently handled. Depending on
+/// whether the handler is invoked for a C++ exception or a failed result<>,
+/// it will have a different type and its `into_result()` method will either
+/// rethrow the exception or return the failed result, respectively.
+///
+/// \section Limitations
+///
+/// The main limitation of result_try is that is is not as flexible with control
+/// flow as a try..catch block is. If you have some logic after the try..catch
+/// block and some code paths in it block return and some don't, you need to
+/// mark it somehow in the return value - for example, you can change result<T>
+/// to result<std::optional<T>> and continue if it is std::nullopt, otherwise
+/// return.
+///
+/// The return type of the first argument must be nothrow move constructible.
+/// Because this function uses many recursive calls to simulate a multi-catch
+/// clause try..catch block and NRVO is not in general guaranteed, there is
+/// a risk that some moves are not elided and one of them can throw
+/// in the middle of the stack. This exception can be thrown after we already
+/// exited the try..catch blocks for some handlers, so not all of them will
+/// participate in handling this exception. Because of the inability to properly
+/// handle exceptions from failed moves, hence the nothrow-move-constructibility
+/// requirement.
+///
+/// This function does not work with futures and will trigger a static_assert
+/// if you use it with a future-returning function. See result_futurize_try
+/// for a version which additionally works with exceptional futures.
+template<typename... Handlers>
+auto result_try(auto&& fun, Handlers&&... handlers) {
```

Comment:
These should most definitely be constrained. Otherwise a compilation error will be a nightmare.

Reply:
Constrained `Fun` and `Handlers` with concepts.